subsys/emds/emds.c

```diff
@@ -189,47 +190,50 @@ int emds_store(void)
 
 	return 0;
 }
+static int entry_load(struct emds_fs *fs, uint16_t id, void *data, size_t entry_len,
+		      enum emds_entry_status *status, bool is_raw)
+{
+	ssize_t len = is_raw ? emds_flash_read_raw(fs, id, data, entry_len, status) :
+			       emds_flash_read(fs, id, data, entry_len, status);
 
-int emds_load(void)
+	if (len <= 0 || len != entry_len) {
+		LOG_ERR("Failed to retrieve dynamic entry: (%d)", id);
+		return -EBADMSG;
+	}
+
+	return 0;
+}
+
+static int stored_load(bool is_raw)
 {
 	struct emds_dynamic_entry *ch;
+	int err = 0;
 
 	if (!emds_initialized) {
 		return -ECANCELED;
 	}
 
 	SYS_SLIST_FOR_EACH_CONTAINER(&emds_dynamic_entries, ch, node) {
-		ssize_t len = emds_flash_read(&emds_flash,
-					      ch->entry.id, ch->entry.data,
-					      ch->entry.len);
+		err |= entry_load(&emds_flash, ch->entry.id, ch->entry.data, ch->entry.len,
+				  &ch->entry.status, is_raw);
+	}
 
-		if (len < 0) {
-			if (len != -ENXIO) {
-				LOG_ERR("Read dynamic entry: (%d) error (%d)",
-					ch->entry.id, len);
-			}
-		} else if (len != ch->entry.len) {
-			LOG_WRN("Read dynamic entry: (%d) did not match (%d:%d).",
-				ch->entry.id, ch->entry.len, len);
-		}
+	STRUCT_SECTION_FOREACH(emds_entry, ch)
+	{
+		err |= entry_load(&emds_flash, ch->id, ch->data, ch->len, &ch->status, is_raw);
 	}
 
-	STRUCT_SECTION_FOREACH(emds_entry, ch) {
-		ssize_t len = emds_flash_read(&emds_flash,
-					      ch->id, ch->data, ch->len);
+	return err;
+}
 
-		if (len < 0) {
-			if (len != -ENXIO) {
-				LOG_ERR("Read static entry: (%d) error (%d)",
-					ch->id, len);
-			}
-		} else if (len != ch->len) {
-			LOG_WRN("Read static entry: (%d) entry did not match (%d:%d)",
-				ch->id, ch->len, len);
-		}
-	}
+int emds_load(void)
+{
+	return stored_load(false);
+}
 
-	return 0;
+int emds_load_raw(void)
+{
+	return stored_load(true);
 }
```

Comment:
Correct, it will load all of the. But assuming that the valid entires is always at the top the same entry will be loaded for all IDs that have a valid entry present and the only change is that the entry missing valid data will get an invalidated one. It creates a bit redundant overhead, but I thought that did not matter as much. I guess with the newly added entry status flag I could improve this a bit.